plugin-e2e/verify-scenario/panelEditPage.spec.ts

```diff
@@ -0,0 +1,85 @@
+import { expect, test } from '@grafana/plugin-e2e';
+
+import { formatExpectError } from './errors';
+import { successfulDataQuery } from './mocks/queries';
+import { scenarios } from './mocks/resources';
+
+const PANEL_TITLE = 'Table panel E2E test';
+const TABLE_VIZ_NAME = 'Table';
+const STANDARD_OTIONS_CATEGORY = 'Standard options';
+const DISPLAY_NAME_LABEL = 'Display name';
+
+test.describe('query editor query data', () => {
+  test('query data response should be OK when query is valid', async ({ panelEditPage }) => {
+    await panelEditPage.datasource.set('gdev-testdata');
+    await expect(
+      panelEditPage.refreshPanel(),
+      formatExpectError('Expected panel query to execute successfully')
+    ).toBeOK();
+  });
+
+  test('query data response should not be OK and panel error should be displayed when query is invalid', async ({
+    panelEditPage,
+  }) => {
+    await panelEditPage.datasource.set('gdev-testdata');
+    const queryEditorRow = await panelEditPage.getQueryEditorRow('A');
+    await queryEditorRow.getByLabel('Labels').fill('invalid-label-format');
+    await expect(panelEditPage.refreshPanel(), formatExpectError('Expected panel query to fail')).not.toBeOK();
+    await expect(
+      panelEditPage,
+      formatExpectError('Expected panel error to be displayed after query execution')
+    ).toHavePanelError();
+  });
+});
+
+test.describe('query editor with mocked responses', () => {
+  test('and resource `scenarios` is mocked', async ({ panelEditPage, selectors }) => {
+    await panelEditPage.mockResourceResponse('scenarios', scenarios);
+    await panelEditPage.datasource.set('gdev-testdata');
```

Comment:
Question: How come we don't have a function directly on the panelEditPage instead of using the datasource property? I'm asking just from a consistency perspective since we have `set` or `get` for most other things directly on the page.

```ts
await panelEditPage.setDatasource('gdev-testdata');
```

Reply:
I've been going for the composable approach here - the DataSourcePicker is a model which is being used across a bunch of different pages. I should be consistent in the sense that the data source is set the same way in the variable edit page, annoation edit page, explore etc.